lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -168,46 +169,39 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
                 args=[], returncode=returncode, stdout=stdout, stderr=None
             )
 
+        TestSendDirectory.function_called.append("fake_tar")
```

Comment:
Yeah, this is one reason I really prefer using instance method test cases instead of static methods; you can just use `self.`. In any case, you could still use `__class__.` ... but while that's a bit more flexible (e.g., if the name were changes later), it's not all that much shorter.

Reply:
Yeah but there are certain mocks like `open` or `unlink` that just take `*args` as a parameter and making them instance method is bit complex. If I do the following, then I can assert that the first argument is a tar path
```
@staticmethod
    def fake_open(*args):
        assert args[0] == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")
``` 
but if I do the following, args become this `('rb',) ` and there is no way to make assertion on the right tar file path
```
@staticmethod
    def fake_open(self, *args):
        # args == ('rb',) 
```

Reply:
@npalaska, your last code blob there is internally inconsistent (at least by convention):  if it's a static method, then there should be no`self` parameter; but, if it's meant to be an instance method, then you need to remove the `@staticmethod` decorator.

By convention, the first parameter to an instance method should be named `self` (and the first parameter to a class method should be named `cls`), but Python doesn't actually care, and you can name it anything you want (but, you should follow the convention...because if you don't, your IDE and your reviewers are likely to get cranky 🙃).  When the function is called, what happens depends totally on what the caller **thinks** the function is:  if it thinks it is a static method, then the first parameter will be bound to the first argument; but, if the caller thinks it is an instance method, then the first parameter will be bound to a reference to the instance, and the second parameter will be bound to the first argument.

So, in last blob there, `self` gets bound to the `Path` instance, and the arguments are passed via `args`.  In your first blob, there's only one parameter, so all the arguments (including the instance ref) are passed in `args`.

So, it doesn't _really_ matter which one you use.  But, for sanity sake, it would be best if the function signature of your mock function properly matched the signature of the function that you're replacing.  In particular, if you are replacing an instance method, your mock should be an instance method -- with the first parameter named `self` and _no_ `@staticmethod` decorator -- and then I think things will follow naturally.

Reply:
@webbnh sorry that was my mistake, thank you for pointing it out but the last blob of the code that I was testing did not have the `staticmethod` decorator it was just (I copied it from the first blob and forgot to remove it)
```
def fake_open(self, *args):
        # args == ('rb',) 
```
and as I said earlier I couldn't assert this statement `args[0] == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")`

Reply:
Yes, that statement doesn't work because, when the mock function is an instance method, the path should be bound to `self` not to `args[0]`.

Reply:
But the `path` is not bound to `self`, `self` is bound to `TestSendDirectory` object so the path is somehow lost.

Reply:
These fixtures are now local to `test_tar_create_success`, so we are out of the woods now w.r.t this complexity.